File: frontend/src/types/queue.ts
export type UserRole = "SUPER_ADMIN" | "HOSPITAL_ADMIN" | "RECEPTIONIST" | "DOCTOR" | "DOCTOR_ASSISTANT";

export type Gender = "MALE" | "FEMALE" | "OTHER" | "UNSPECIFIED";

export type QueueStatus = "OPEN" | "PAUSED" | "CLOSED";

export type TokenStatus =
  | "WAITING"
  | "READY"
  | "AWAY"
  | "RETURNING"
  | "CALLED"
  | "SERVING"
  | "MISSED"
  | "SKIPPED"
  | "COMPLETED"
  | "CANCELLED"
  | "TRANSFERRED";

export type PriorityLevel = "EMERGENCY" | "HIGH" | "NORMAL";

export interface StaffUser {
  id: string;
  hospital_id: string;
  branch_id?: string;
  email: string;
  full_name: string;
  phone_number?: string;
  role: UserRole;
  is_active: boolean;
  created_at: string;
}

export interface Hospital {
  id: string;
  name: string;
  slug: string;
  address?: string;
  phone?: string;
  is_active: boolean;
  created_at: string;
}

export interface Department {
  id: string;
  branch_id: string;
  name: string;
  code: string;
  is_active: boolean;
}

export interface Queue {
  id: string;
  department_id: string;
  doctor_user_id?: string;
  room_id?: string;
  name: string;
  prefix: string;
  status: QueueStatus;
  default_consult_time_min: number;
  current_sequence: number;
  rejoin_policy: {
    strategy: string;
    offset: number;
    max_rejoins: number;
  };
  created_at: string;
}

export interface QueueToken {
  id: string;
  public_id: string;
  queue_id: string;
  visit_id: string;
  patient_id: string;
  token_display_number: string;
  sequence_number: number;
  priority: PriorityLevel;
  status: TokenStatus;
  operational_position?: number;
  estimated_wait_min?: number;
  estimated_wait_max?: number;
  missed_count: number;
  rejoin_count: number;
  created_at: string;
  ready_at?: string;
  called_at?: string;
  serving_at?: string;
  completed_at?: string;
}

export interface PatientLiveTokenView {
  public_id: string;
  token_display_number: string;
  sequence_number: number;
  status: TokenStatus;
  priority: PriorityLevel;
  queue_id: string;
  queue_name: string;
  queue_status: QueueStatus;
  doctor_name?: string;
  department_name: string;
  room_number?: string;
  currently_serving_token_number?: string;
  currently_called_token_number?: string;
  patients_ahead: number;
  operational_position?: number;
  estimated_wait_min?: number;
  estimated_wait_max?: number;
  estimated_wait_display: string;
  action_prompt: string;
  can_mark_away: boolean;
  can_mark_returning: boolean;
  can_mark_ready: boolean;
  created_at: string;
  ready_at?: string;
  called_at?: string;
  serving_at?: string;
  completed_at?: string;
}

export interface QueueSummary {
  queue: Queue;
  currently_serving_token?: QueueToken;
  currently_called_token?: QueueToken;
  total_waiting: number;
  total_ready: number;
  total_away: number;
  total_completed_today: number;
  active_tokens: QueueToken[];
}

File: backend/app/schemas/queue.py
import uuid
from datetime import datetime
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, ConfigDict, Field
from app.models.enums import QueueStatus, TokenStatus, PriorityLevel, Gender


class QueueBase(BaseModel):
    name: str = Field(..., min_length=2, max_length=100)
    prefix: str = Field("A", min_length=1, max_length=10)
    default_consult_time_min: int = Field(10, ge=1, le=120)
    rejoin_policy: Dict[str, Any] = Field(
        default_factory=lambda: {
            "strategy": "OFFSET_BEHIND_CURRENT",
            "offset": 2,
            "max_rejoins": 2,
        }
    )


class QueueCreate(QueueBase):
    department_id: uuid.UUID
    doctor_user_id: Optional[uuid.UUID] = None
    room_id: Optional[uuid.UUID] = None


class QueueUpdate(BaseModel):
    name: Optional[str] = None
    prefix: Optional[str] = None
    doctor_user_id: Optional[uuid.UUID] = None
    room_id: Optional[uuid.UUID] = None
    default_consult_time_min: Optional[int] = None
    rejoin_policy: Optional[Dict[str, Any]] = None
    status: Optional[QueueStatus] = None


class QueueOut(QueueBase):
    id: uuid.UUID
    department_id: uuid.UUID
    doctor_user_id: Optional[uuid.UUID] = None
    room_id: Optional[uuid.UUID] = None
    status: QueueStatus
    current_sequence: int
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class QueueTokenOut(BaseModel):
    id: uuid.UUID
    public_id: str
    queue_id: uuid.UUID
    visit_id: uuid.UUID
    patient_id: uuid.UUID
    token_display_number: str
    sequence_number: int
    priority: PriorityLevel
    status: TokenStatus
    operational_position: Optional[int] = None
    estimated_wait_min: Optional[int] = None
    estimated_wait_max: Optional[int] = None
    missed_count: int
    rejoin_count: int
    created_at: datetime
    ready_at: Optional[datetime] = None
    called_at: Optional[datetime] = None
    serving_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


class QueueTokenCreateWalkIn(BaseModel):
    queue_id: uuid.UUID
    # Patient Information (auto finds or creates patient)
    patient_name: str = Field(..., min_length=2, max_length=255)
    patient_phone: Optional[str] = Field(None, max_length=50)
    patient_gender: Gender = Gender.UNSPECIFIED
    # Consultation details
    priority: PriorityLevel = PriorityLevel.NORMAL
    notes: Optional[str] = None


class QueuePauseRequest(BaseModel):
    reason: str = Field(..., min_length=2, max_length=255)
    expected_resume_minutes: Optional[int] = Field(None, ge=1, le=240)


class QueueSummaryOut(BaseModel):
    queue: QueueOut
    currently_serving_token: Optional[QueueTokenOut] = None
    currently_called_token: Optional[QueueTokenOut] = None
    total_waiting: int
    total_ready: int
    total_away: int
    total_completed_today: int
    active_tokens: List[QueueTokenOut]
